File: bindings/python/src/zlink/contracts/sockets/pubsub_socket_contracts.py
# SPDX-License-Identifier: MPL-2.0

from typing import Protocol, runtime_checkable

from . import socket as _socket_contract


@runtime_checkable
class PubSocket(_socket_contract._SocketContract, Protocol):
    """PUB socket: topic-filtered publish that drops messages with no matching
    subscriber."""

    @property
    def pub_options(self):
        """The PUB-specific typed options facade."""
        ...

    def connect(self, endpoint):
        """Connect to a remote transport address (asynchronous)."""
        ...

    def disconnect(self, endpoint):
        """Disconnect the connection previously established to ``endpoint``."""
        ...

    def publish(self, topic):
        """Begin publishing under ``topic``; parts are consumed on a successful
        submit."""
        ...


@runtime_checkable
class SubSocket(_socket_contract._SocketContract, Protocol):
    """SUB socket: a receive-only subscriber filtered by its subscriptions."""

    @property
    def sub_options(self):
        """The SUB-specific typed options facade."""
        ...

    def connect(self, endpoint):
        """Connect to a remote transport address (asynchronous)."""
        ...

    def disconnect(self, endpoint):
        """Disconnect the connection previously established to ``endpoint``."""
        ...

    def set_subscription(self, topic):
        """Add a subscription for ``topic`` (an exact topic or pattern);
        subscriptions accumulate."""
        ...

    def unset_subscription(self, topic):
        """Remove a subscription previously added for ``topic``."""
        ...

    def subscription_at(self, index):
        """Return ``(filter, is_pattern)`` for a subscription index, or
        ``None`` when the index is absent."""
        ...

    def subscribe_into(self, topic_message, *, flags=0):
        """Receive the next matching topic message into ``topic_message``
        storage; ``False`` when ``DONT_WAIT`` is set and none is available."""
        ...


@runtime_checkable
class XPubSocket(_socket_contract._SocketContract, Protocol):
    """XPUB socket: like PUB, but also surfaces subscriber subscription and
    unsubscription events."""

    @property
    def pub_options(self):
        """The PUB/XPUB-specific typed options facade."""
        ...

    def connect(self, endpoint):
        """Connect to a remote transport address (asynchronous)."""
        ...

    def disconnect(self, endpoint):
        """Disconnect the connection previously established to ``endpoint``."""
        ...

    def publish(self, topic):
        """Begin publishing under ``topic``; parts are consumed on a successful
        submit."""
        ...

    def receive_subscription_event_into(self, event, *, flags=0):
        """Receive the next subscriber (un)subscription event into ``event``
        storage; ``False`` when ``DONT_WAIT`` is set and none is available."""
        ...


@runtime_checkable
class XSubSocket(_socket_contract._SocketContract, Protocol):
    """XSUB socket: a subscriber whose subscriptions are carried as messages."""

    @property
    def sub_options(self):
        """The SUB-specific typed options facade."""
        ...

    def connect(self, endpoint):
        """Connect to a remote transport address (asynchronous)."""
        ...

    def disconnect(self, endpoint):
        """Disconnect the connection previously established to ``endpoint``."""
        ...

    def set_subscription(self, topic):
        """Add a subscription for ``topic``; subscriptions accumulate."""
        ...

    def unset_subscription(self, topic):
        """Remove a subscription previously added for ``topic``."""
        ...

    def subscription_at(self, index):
        """Return ``(filter, is_pattern)`` for a subscription index, or
        ``None`` when the index is absent."""
        ...

    def subscribe_into(self, topic_message, *, flags=0):
        """Receive the next matching topic message into ``topic_message``
        storage; ``False`` when ``DONT_WAIT`` is set and none is available."""
        ...


__all__ = ["PubSocket", "SubSocket", "XPubSocket", "XSubSocket"]

File: bindings/dotnet/src/Zlink/Contracts/Sockets/PubSubSocketContracts.cs
// SPDX-License-Identifier: MPL-2.0

namespace Systems.Zlink;

/// <summary>
///     Contract for sockets that publish messages under a topic.
/// </summary>
public interface IPublisherSocket : IConnectableSocket
{
    /// <summary>
    ///     Begins publishing under <paramref name="topic" />. Publish is
    ///     synchronous: PUB semantics are lossy, so the publisher never waits at
    ///     the high-water mark and the terminal
    ///     <see cref="PublishSubmitOperation.Submit" /> completes on the calling
    ///     thread. With <c>NODROP</c> a full subscriber surfaces immediately as
    ///     <see cref="ZlinkSubmitException" />; the retry policy belongs to the
    ///     application.
    /// </summary>
    PublishOperation Publish(string topic);

    /// <summary>
    ///     Begins an explicit immediate publish attempt. Add parts and call
    ///     <see cref="SendSubmitOperation.Submit" />; use
    ///     <see cref="SendFlags.DontWait" /> to observe back-pressure as a
    ///     <c>false</c> result rather than an exception.
    /// </summary>
    TryPublishOperation TryPublish(string topic);
}

/// <summary>
///     Contract for sockets that subscribe to published topics.
/// </summary>
public interface ISubscriberSocket : IConnectableSocket
{
    /// <summary>
    ///     Adds a subscription for <paramref name="topicOrPattern" />. Subscriptions
    ///     accumulate; a socket may hold several at once.
    /// </summary>
    void SetSubscription(string topicOrPattern);

    /// <summary>
    ///     Removes a subscription previously added for
    ///     <paramref name="topicOrPattern" />.
    /// </summary>
    void UnsetSubscription(string topicOrPattern);

    /// <summary>
    ///     Gets the active subscription at <paramref name="index" />, or null when
    ///     the index is out of range.
    /// </summary>
    SubscriptionEntry? SubscriptionAt(int index);

    /// <summary>
    ///     Receives the next matching topic message into caller-provided storage.
    /// </summary>
    /// <param name="result">Reusable storage overwritten on success.</param>
    /// <param name="flags">Receive behavior flags.</param>
    /// <returns>
    ///     true on success; false when <see cref="RecvFlags.DontWait" /> is set and
    ///     no message is available.
    /// </returns>
    bool Subscribe(TopicMessage result, RecvFlags flags = RecvFlags.None);
}

/// <summary>
///     Contract for a PUB socket: topic-filtered publish that drops messages with
///     no matching subscriber.
/// </summary>
public interface IPubSocket : IPublisherSocket
{
    /// <summary>
    ///     Gets the PUB-specific typed options facade.
    /// </summary>
    new PubSocketOptions Options { get; }

    /// <summary>
    ///     Sets the routing id used to identify this publisher socket. Apply before
    ///     binding or connecting so peers observe the configured identity.
    /// </summary>
    void SetRoutingId(RoutingId routingId);

    /// <summary>
    ///     Gets the routing id currently assigned to this publisher socket.
    /// </summary>
    RoutingId GetRoutingId();

}

/// <summary>
///     Contract for a SUB socket: receive-only subscriber filtered by its
///     subscriptions.
/// </summary>
public interface ISubSocket : ISubscriberSocket
{
    /// <summary>
    ///     Gets the SUB-specific typed options facade.
    /// </summary>
    new SubSocketOptions Options { get; }

    /// <summary>
    ///     Sets the routing id used to identify this subscriber socket. Apply before
    ///     binding or connecting so peers observe the configured identity.
    /// </summary>
    void SetRoutingId(RoutingId routingId);

    /// <summary>
    ///     Gets the routing id currently assigned to this subscriber socket.
    /// </summary>
    RoutingId GetRoutingId();

}

/// <summary>
///     Contract for an XPUB socket: like PUB, but also surfaces subscriber
///     subscription and unsubscription events via
///     <see cref="ReceiveSubscriptionEvent" />.
/// </summary>
public interface IXPubSocket : IPublisherSocket
{
    /// <summary>
    ///     Gets the PUB-specific typed options facade.
    /// </summary>
    new PubSocketOptions Options { get; }

    /// <summary>
    ///     Receives the next subscriber (un)subscription event into
    ///     caller-provided storage.
    /// </summary>
    /// <param name="result">Reusable storage overwritten on success.</param>
    /// <param name="flags">Receive behavior flags.</param>
    /// <returns>
    ///     true on success; false when <see cref="RecvFlags.DontWait" /> is set and
    ///     no event is available.
    /// </returns>
    bool ReceiveSubscriptionEvent(SubscriptionEvent result,
        RecvFlags flags = RecvFlags.None);
}

/// <summary>
///     Contract for an XSUB socket: a subscriber whose subscriptions are carried as
///     messages rather than set as socket options.
/// </summary>
public interface IXSubSocket : ISubscriberSocket
{
    /// <summary>
    ///     Gets the SUB-specific typed options facade.
    /// </summary>
    new SubSocketOptions Options { get; }
}
